presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractArrayBlock.java

```diff
@@ -115,6 +115,11 @@ public Block copyRegion(int position, int length)
         int endValueOffset = getOffset(position + length);
         Block newValues = getValues().copyRegion(startValueOffset, endValueOffset - startValueOffset);
 
+        if (position == 0 && length == positionCount &&
+                getOffsetBase() == 0 && getOffsets().length == positionCount + 1 && getValueIsNull().length == positionCount) {
+            return new ArrayBlock(length, getValueIsNull(), getOffsets(), newValues);
```

Comment:
This makes we always create a new `ArrayBlock`, so it's difficult to tell whether the original block is compact or not, and thus makes it difficult to monitor stats / create test.

We can return `this` by changing this line to:
```
   if (newValues == getValues()) {
      return this;
   }
   return new ArrayBlock(length, getValueIsNull(), getOffsets(), newValues);
```

An alternative way would be to add `compact()` and `isCompact()` into the Block interface. Today `DictionaryBlock` already has these two interfaces.

This situation also applies to `MapBlock` and `RowBlock`.